nmoscommon/auth/nmos_auth.py

```diff
@@ -42,83 +43,109 @@
+                    return cert
         except OSError:
-            self.logger.writeError("File does not exist or you do not have permission to open it")
+            self.logger.writeError("File {} does not exist or you do not have permission to open it".format(
+                filename))
             raise
 
-    def extractPublicKey(self, certificate):
-        crtObj = crypto.load_certificate(crypto.FILETYPE_PEM, certificate)
-        pubKeyObject = crtObj.get_pubkey()
-        pubKeyString = crypto.dump_publickey(crypto.FILETYPE_PEM, pubKeyObject)
-        if pubKeyString is None:
-            self.logger.writeError("Public Key could not be extracted from certificate")
-            raise ValueError
-        else:
-            return pubKeyString.decode('utf-8')
+    def getPublicKeyString(self, key_class):
+        serialised_key = key_class.public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo
+        )
+        return serialised_key.decode('utf-8')
+
+    def extractPublicKey(self, key_containing_object):
+        """Extracts a key from the given parameter. A list or a dict object will be treated as a JWK structure.
+        A string will be treated like a X509 certificate"""
+        try:
+            if isinstance(key_containing_object, dict):
+                key_class = jwk.loads(key_containing_object)
+                return self.getPublicKeyString(key_class)
+            elif isinstance(key_containing_object, list):
+                serialised_keys = []
+                for key in key_containing_object:
+                    key_class = jwk.loads(key)
+                    serialised_keys.append(
+                        self.getPublicKeyString(key_class)
+                    )
+                return serialised_keys[0]  # TODO - manage priority of keys when format is agreed
+            elif isinstance(key_containing_object, str):
+                crt_obj = x509.load_pem_x509_certificate(key_containing_object, default_backend())
+                key_class = crt_obj.public_key()
+                pubkey_string = self.getPublicKeyString(key_class)
+                if pubkey_string is None:
+                    self.logger.writeError("Public Key could not be extracted from certificate")
+                else:
+                    return pubkey_string
+        except Exception:
+            self.logger.writeError("Public Key(s) could not be extracted from JSON Web Keys")
+            raise
 
     def getPublicKey(self):
-        if self.certificate is None:
-            self.logger.writeInfo("Fetching Certificate...")
+        if self.publicKey is None:
+            self.logger.writeInfo("Fetching JSON Web Keys...")
             try:
-                self.logger.writeInfo("Trying to fetch cert using mDNS...")
-                cert = self.getCertFromEndpoint()
+                self.logger.writeInfo("Trying to fetch keys using mDNS...")
```

Comment:
This message may be a little misleading as we could use unicast DNS too, and the keys themselves are fetched over HTTP. Perhaps 'Trying to discover keys using DNS-SD...' or similar?